### PR TITLE
Add context menu with option to jump to Caller/Callee

### LINKED ISSUE #33: jump to callers/callee view from stack views
The stack views, i.e. bottom-up, top-down and flame-graph should get context menus that show the selected symbol in the caller/callee view. I.e. set the current index, and ensure it's visible.

### PR DESCRIPTION
**Pull request by nwrogers**
Add context menu with option to jump to Caller/Callee tab from Bottom-Up, Top-Down, and Flame-Graph.

Fixes #33

### CODE REVIEW

File: src/mainwindow.cpp [hunk 1]
```diff
@@ -367,3 +375,36 @@ void MainWindow::aboutHotspot()
     dialog.adjustSize();
     dialog.exec();
 }
+
+void MainWindow::customContextMenu(const QPoint &point, QTreeView* view, quint32 symbolRole)
```
Comment (milianw) [inline src/mainwindow.cpp]: use int instead of quint32

File: src/mainwindow.cpp [hunk 2]
```diff
@@ -367,3 +375,36 @@ void MainWindow::aboutHotspot()
...
+{
+    QModelIndex index = view->indexAt(point);
+    if (index.isValid()) {
```
Comment (milianw) [inline src/mainwindow.cpp]: invert and do early-return when it's not valid, decreases the indentation depth below

File: src/mainwindow.cpp [hunk 3]
```diff
@@ -367,3 +375,36 @@ void MainWindow::aboutHotspot()
...
+        QMenu contextMenu;
+        QAction *invokeAction = nullptr;
```
Comment (milianw) [inline src/mainwindow.cpp]: merge with next line (see flamegraph comment above)

Comment (milianw) [inline src/mainwindow.cpp]: also rename

File: src/mainwindow.cpp [hunk 4]
```diff
@@ -367,3 +375,36 @@ void MainWindow::aboutHotspot()
...
+{
+    customContextMenu(point, ui->bottomUpTreeView, BottomUpModel::SymbolRole);
+}
+
+void MainWindow::onTopDownContextMenu(const QPoint &point)
+{
+    customContextMenu(point, ui->topDownTreeView, TopDownModel::SymbolRole);
+}
+
+void MainWindow::onJumpToCallerCallee(const Data::Symbol& symbol)
```
Comment (milianw) [inline src/mainwindow.cpp]: remove the on prefix

Comment (milianw) [inline src/mainwindow.cpp]: move & next to arg name instead of type name

File: src/mainwindow.h [hunk 5]
```diff
@@ -30,6 +30,11 @@
 #include <QMainWindow>
 #include <QScopedPointer>
 #include <QString>
+#include <QSortFilterProxyModel>
+#include <QTreeView>
+
+#include "callercalleemodel.h"
```
Comment (milianw) [inline src/mainwindow.h]: foward-declare the model and the symbol, that should work - no need to include the fat headers here

File: src/mainwindow.h [hunk 6]
```diff
@@ -30,6 +30,11 @@
 #include <QMainWindow>
 #include <QScopedPointer>
 #include <QString>
+#include <QSortFilterProxyModel>
```
Comment (milianw) [inline src/mainwindow.h]: forward-declare both

File: src/models/hashmodel.h [hunk 7]
```diff
@@ -104,6 +106,16 @@ class HashModel : public QAbstractTableModel
         return index(row, column);
     }
 
+    QModelIndex indexForSymbol(Data::Symbol symbol)
```
Comment (milianw) [inline src/models/hashmodel.h]: const&

Comment (milianw) [inline src/models/hashmodel.h]: this method should be moved into the SymbolHashModel, in `callercalleemodel.h`

File: src/mainwindow.cpp [hunk 8]
```diff
@@ -367,3 +375,36 @@ void MainWindow::aboutHotspot()
...
+{
+    QModelIndex index = view->indexAt(point);
+    if (index.isValid()) {
+        QMenu contextMenu;
+        QAction *invokeAction = nullptr;
+        invokeAction = contextMenu.addAction(tr("View Caller/Callee"));
+
+        QAction *action = contextMenu.exec(QCursor::pos());
+        if (action == invokeAction) {
+            const auto symbol = index.data(symbolRole).template value<Data::Symbol>();
```
Comment (milianw) [inline src/mainwindow.cpp]: the `template ` part shouldn't be required here, just `value` should be enough

File: src/mainwindow.cpp [hunk 9]
```diff
@@ -367,3 +375,36 @@ void MainWindow::aboutHotspot()
...
+
+void MainWindow::customContextMenu(const QPoint &point, QTreeView* view, quint32 symbolRole)
+{
+    QModelIndex index = view->indexAt(point);
+    if (index.isValid()) {
+        QMenu contextMenu;
+        QAction *invokeAction = nullptr;
+        invokeAction = contextMenu.addAction(tr("View Caller/Callee"));
+
+        QAction *action = contextMenu.exec(QCursor::pos());
```
Comment (milianw) [inline src/mainwindow.cpp]: auto *

File: src/mainwindow.cpp [hunk 10]
```diff
@@ -367,3 +375,36 @@ void MainWindow::aboutHotspot()
     dialog.adjustSize();
     dialog.exec();
 }
+
+void MainWindow::customContextMenu(const QPoint &point, QTreeView* view, quint32 symbolRole)
+{
+    QModelIndex index = view->indexAt(point);
```
Comment (milianw) [inline src/mainwindow.cpp]: const auto

File: src/flamegraph.h [hunk 11]
```diff
@@ -80,6 +83,7 @@ private slots:
     bool m_collapseRecursion = false;
     // cost threshold in percent, items below that value will not be shown
     double m_costThreshold = 0.1;
+    const FrameGraphicsItem* m_contextMenuItem = nullptr;
```
Comment (milianw) [inline src/flamegraph.h]: remove once you cleaned up the event handler

File: src/models/hashmodel.h (hunk 7 above)
Comment (nwrogers) [inline src/models/hashmodel.h]: m_rows is private, is there another way to get the rows from outside the HashModel without having to make m_rows protected? Should I make m_rows protected? Also what is SymbolHashModel?

File: src/mainwindow.cpp [hunk 12]
```diff
@@ -214,10 +218,12 @@ MainWindow::MainWindow(QWidget *parent) :
             });
 
     auto calleesModel = setupCallerOrCalleeView<CalleeModel>(ui->calleesView, ui->callerCalleeTableView,
-                                                             callerCalleeCostModel, callerCalleeProxy);
+                                                             m_callerCalleeCostModel, m_callerCalleeProxy);
+    auto calleeCostDelegate = new CostDelegate(CalleeModel::SortRole, CalleeModel::TotalCostRole, this);
```
Comment (nwrogers) [inline src/mainwindow.cpp]: Messed up the merge :/

File: src/flamegraph.cpp [hunk 13]
```diff
@@ -447,6 +448,24 @@ bool FlameGraph::eventFilter(QObject* object, QEvent* event)
         updateTooltip();
     } else if (event->type() == QEvent::Hide) {
         setData(nullptr);
+    } else if (event->type() == QEvent::MouseButtonPress) {
```
Comment (nwrogers) [inline src/flamegraph.cpp]: I originally tried to do it all in event->type() == QEvent::ContextMenu but When I tried to get the FrameGraphicsItem I always got NULL. I checked the position returned from mapFromGlobal and it was off by one pixel in x and one pixel in y, so I made the one pixel adjustment and voila! I get our item... not sure what is worse the old way or the new way...

QPoint itemPosition(QCursor::pos().x()-1, QCursor::pos().y()-1);
auto item = static_cast<FrameGraphicsItem*>(m_view->itemAt(m_view->mapFromGlobal(itemPosition)));

Comment (milianw) [inline src/flamegraph.cpp]: odd, could you try http://doc.qt.io/qt-5/qcontextmenuevent.html#globalPos - does that work?

File: src/models/hashmodel.h (hunk 7 above)
Comment (milianw) [inline src/models/hashmodel.h]: sorry, I mean `SymbolCostModelImpl`, and yes make m_rows protected for that purpose

Comment (nwrogers) [inline src/models/hashmodel.h]: Did you mean I should move it to CallerCalleeModel? I don't call indexForSymbol on the CostModel, at least not at this time

File: src/flamegraph.cpp (hunk 13 above)
Comment (nwrogers) [inline src/flamegraph.cpp]: That did the trick

File: src/flamegraph.cpp [hunk 14]
```diff
@@ -447,6 +448,16 @@ bool FlameGraph::eventFilter(QObject* object, QEvent* event)
         updateTooltip();
     } else if (event->type() == QEvent::Hide) {
         setData(nullptr);
+    } else if (event->type() == QEvent::ContextMenu) {
+        QMenu contextMenu;
+        auto *viewCallerCallee = contextMenu.addAction(tr("View Caller/Callee"));
+
+        QContextMenuEvent *contextEvent = static_cast<QContextMenuEvent*>(event);
+        auto item = static_cast<FrameGraphicsItem*>(m_view->itemAt(m_view->mapFromGlobal(contextEvent->globalPos())));
```
Comment (milianw) [inline src/flamegraph.cpp]: actually, when no item is found, we shouldn't even offer the context menu, right? so please change this to check for an item first before creating the menu and action

File: src/mainwindow.cpp [hunk 15]
```diff
@@ -367,3 +373,36 @@ void MainWindow::aboutHotspot()
     dialog.adjustSize();
     dialog.exec();
 }
+
+void MainWindow::customContextMenu(const QPoint &point, QTreeView* view, int symbolRole)
+{
+    const auto index = view->indexAt(point);
+    if (!index.isValid()) {
```
Comment (milianw) [inline src/mainwindow.cpp]: this is the correct way of doing it, the flamegraph should do the same (see comment above)

**Comment by milianw** [comment]
oh, you'll need to resolve the conflicts and resubmit - I added some new API in mainwindow.h

**Pull request merged**